Feature/796 improved health check design

lewisdaly opened the pull request:
Part of: https://github.com/mojaloop/project/issues/796

This PR contains the design and sequence diagrams for an improved health check. For now, we are documenting this on the `central-ledger` only, but we will apply this design to the following services:

- ml-api-adapter
- central-ledger
- central-settlement
- central-event-processor
- email-notifier

The Rendered design document can be found here: https://github.com/vessels-tech/documentation/blob/feature/796-improved-health-check-design/mojaloop-technical-overview/central-ledger/admin-operations/1.0.0-get-health-check.md

>_Note: This is my first sequence diagram and api design for Mojaloop, so please let me know what I'm doing that can be improved!_

mdebarros commented on mojaloop-technical-overview/central-ledger/admin-operations/1.0.0-get-health-check.md:
```diff
@@ -0,0 +1,223 @@
+| --- | --- | --- | --- |
+| `name`    | `enum`   | The service name. Options are `kafka`, `mysql` and `sidecar`. | `"kafka` | 
+| `status`  | `enum`   | The status of the service. Options are `OK` and `DOWN`        | `"OK"` | 
+| `latenct` | `number` | Latency (ms) when connecting to the service.                  | `459"` | 
```
Typo `latenct` --> `latency`

mdebarros commented on mojaloop-technical-overview/central-ledger/admin-operations/1.0.0-get-health-check.md:
```diff
@@ -0,0 +1,223 @@
+
+| Name  | Type | Description | Example |
+| --- | --- | --- | --- |
+| `status`        | `enum`                             | The status of the service. Options are `OK` and `DOWN` | `"OK"`                       | 
```
As discussed.

Please clarify the meaning of `OK` vs `DOWN`. Aka `OK` means that the API is functioning correctly, and all back-end serviceHealth checks have passed, etc.

You could also consider an intermediate level indicating that the API is up but that back-end serviceHealth checks have failed. I feel like this may be more complex then we need it to be, and may introduce some confusion. Let me know what you think.

lewisdaly replied:
I'd prefer to keep it as a binary setting for now to keep things simple. If we want to check to see if the API is healthy, but the sub-services are not, then we can simply look for the `502 status code`

lewisdaly commented:
Don't merge quite yet. Still discussing some details in the [issue](https://github.com/mojaloop/project/issues/796).

lewisdaly commented:
@mdebarros feel free to take another look - I've updated the design to remove latency and the sidecar. I also added some more clarity around the definition of `OK` and `DOWN` for sub-services.

mdebarros commented on mojaloop-technical-overview/central-ledger/admin-operations/1.0.0-get-health-check.md:
```diff
@@ -59,6 +53,33 @@ When a service is `OK`: the API is considered healthy, and all sub-services are
+- An existing connection to the kafka broker
+- The necessary topics exist. This will change depending on which service the health check is running for.
+
+For example, for the `central-ledger` service to be considered healthy, the following topics need to be found:
```
Would this be necessary considering we are creating consumers for each of the topics?

Aka, would the consumer not fail to start up if the admin topic did not exist?

lewisdaly replied:
Yes, true, and the same is the case for the database as well.

I might leave @elnyry to answer this one in more depth, but from another discussion I think it's also about setting up a better health checking methodology that will feed into improve our ability to scale and make services inside of k8s more reliable and fault tolerant in the future. 

I think this may also change as we introduce other sub-services as well - but I do understand what you are getting at. With this current design, the health check would never fail since the api would never start if it can't find the admin topic or connect to the api.

mdebarros commented on mojaloop-technical-overview/central-ledger/admin-operations/1.0.0-get-health-check.md:
```diff
@@ -43,9 +38,8 @@ Responses Codes:
 | Name  | Type | Description | Example |
 | --- | --- | --- | --- |
-| `name`    | `enum`   | The service name. Options are `kafka`, `mysql` and `sidecar`. | `"kafka` | 
+| `name`    | `enum`   | The service name. Options are `kafka` and `mysql`             | `"kafka` | 
```
I think you may also need to add "sidecar".

That is very likely going to be required for the event framework that is going to be proposed shortly.

lewisdaly replied:
Yeah I like the idea of generalizing the enums - I was just thinking that `mysql` is rather opinionated. 

We can also leave sidecar in the design for now but perhaps not enable is during the refactor - that way we won't need to revise the design. I was just leaving it out as @elnyry 's recommendation.
